// @volar/vue-language-server/out/registers/registerlanguageFeatures.js
Object.defineProperty(exports, "__esModule", { value: true });
exports.register = void 0;
const vue = require("@volar/vue-language-service");
const vscode = require("vscode-languageserver");
function register(features, legend, server, languageConfigs) {
    var _a, _b, _c, _d;
    if (features.references) {
        server.referencesProvider = true;
    }
    if (features.implementation) {
        server.implementationProvider = true;
    }
    if (features.definition) {
        server.definitionProvider = true;
    }
    if (features.typeDefinition) {
        server.typeDefinitionProvider = true;
    }
    if (features.callHierarchy) {
        server.callHierarchyProvider = true;
    }
    if (features.hover) {
        server.hoverProvider = true;
    }
    if (features.rename) {
        server.renameProvider = {
            prepareProvider: true,
        };
    }
    if (features.renameFileRefactoring) {
        server.workspace = {
            fileOperations: {
                willRename: {
                    filters: [
                        ...[...languageConfigs.definitelyExts, ...languageConfigs.indeterminateExts].map(ext => ({ pattern: { glob: `**/*${ext}` } })),
                        { pattern: { glob: '**/*.js' } },
                        { pattern: { glob: '**/*.ts' } },
                        { pattern: { glob: '**/*.jsx' } },
                        { pattern: { glob: '**/*.tsx' } },
                        { pattern: { glob: '**/*.json' } },
                    ]
                }
            }
        };
    }
    if (features.signatureHelp) {
        server.signatureHelpProvider = {
            triggerCharacters: ['(', ',', '<'],
            retriggerCharacters: [')'],
        };
    }
    if (features.completion) {
        server.completionProvider = {
            triggerCharacters: '!@#$%^&*()_+-=`~{}|[]\:";\'<>?,./ '.split(''),
            resolveProvider: true,
        };
        server.executeCommandProvider = {
            commands: [
                ...((_b = (_a = server.executeCommandProvider) === null || _a === void 0 ? void 0 : _a.commands) !== null && _b !== void 0 ? _b : []),
                'volar.server.convertTagNameCasing',
            ]
        };
    }
    if (features.documentHighlight) {
        server.documentHighlightProvider = true;
    }
    if (features.documentLink) {
        server.documentLinkProvider = {
            resolveProvider: false, // TODO
        };
    }
    if (features.workspaceSymbol) {
        server.workspaceSymbolProvider = true;
    }
    if (features.codeLens) {
        server.codeLensProvider = {
            resolveProvider: true,
        };
        server.executeCommandProvider = {
            commands: [
                ...((_d = (_c = server.executeCommandProvider) === null || _c === void 0 ? void 0 : _c.commands) !== null && _d !== void 0 ? _d : []),
                vue.executePluginCommand,
            ]
        };
    }
    if (features.semanticTokens) {
        server.semanticTokensProvider = {
            range: true,
            full: false,
            legend,
        };
    }
    if (features.codeAction) {
        server.codeActionProvider = {
            codeActionKinds: [
                vscode.CodeActionKind.Empty,
                vscode.CodeActionKind.QuickFix,
                vscode.CodeActionKind.Refactor,
                vscode.CodeActionKind.RefactorExtract,
                vscode.CodeActionKind.RefactorInline,
                vscode.CodeActionKind.RefactorRewrite,
                vscode.CodeActionKind.Source,
                vscode.CodeActionKind.SourceFixAll,
                vscode.CodeActionKind.SourceOrganizeImports,
            ],
            resolveProvider: true,
        };
    }
    if (features.inlayHints) {
        server.inlayHintProvider = true;
    }
}
exports.register = register;
