// @volar/vue-language-service/out/utils/sharedLs.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getDummyTsLs = void 0;
const shared = require("@volar/shared");
// Fast dummy TS language service, only has one script.
let dummyProjectVersion = 0;
let dummyTsLs;
let doc;
function getDummyTsLs(ts, ts2, _doc, settings) {
    if (!dummyTsLs) {
        const host = {
            readFile: ts.sys.readFile,
            fileExists: ts.sys.fileExists,
            getProjectVersion: () => dummyProjectVersion.toString(),
            getScriptVersion: () => dummyProjectVersion.toString(),
            getCompilationSettings: () => ({ allowJs: true, jsx: ts.JsxEmit.Preserve }),
            getScriptFileNames: () => [shared.uriToFsPath(doc.uri)],
            getScriptSnapshot: fileName => {
                if (shared.fsPathToUri(fileName) === shared.normalizeUri(doc.uri)) {
                    return ts.ScriptSnapshot.fromString(doc.getText());
                }
            },
            getCurrentDirectory: () => '',
            getDefaultLibFileName: () => '',
        };
        dummyTsLs = ts2.createLanguageService(ts, host, ts.createLanguageService(host), settings);
    }
    dummyProjectVersion++;
    doc = _doc;
    return dummyTsLs;
}
exports.getDummyTsLs = getDummyTsLs;
