// @volar/vue-language-server/out/browser.js
Object.defineProperty(exports, "__esModule", { value: true });
const vscode = require("vscode-languageserver/browser");
const common_1 = require("./common");
const request_light_1 = require("request-light");
const http_1 = require("./schemaRequestHandlers/http");
const ts = require("typescript/lib/tsserverlibrary"); // bundle typescript lib in web
const messageReader = new vscode.BrowserMessageReader(self);
const messageWriter = new vscode.BrowserMessageWriter(self);
const connection = vscode.createConnection(messageReader, messageWriter);
(0, common_1.createLanguageServer)(connection, {
    loadTypescript(options) {
        return ts; // not support load by user config in web
    },
    loadTypescriptLocalized(options) {
        // TODO
    },
    schemaRequestHandlers: {
        http: http_1.default,
        https: http_1.default,
    },
    onDidChangeConfiguration(settings) {
        (0, request_light_1.configure)(settings.http && settings.http.proxy, settings.http && settings.http.proxyStrictSSL);
    },
    fileSystemProvide: undefined, // TODO
});
