// @volar/vue-language-service/out/languageFeatures/codeActions.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.register = void 0;
const shared = require("@volar/shared");
const transforms_1 = require("@volar/transforms");
const vscode = require("vscode-languageserver-protocol");
const dedupe = require("../utils/dedupe");
const featureWorkers_1 = require("../utils/featureWorkers");
const rename_1 = require("./rename");
function register(context) {
    return (uri, range, codeActionContext) => __awaiter(this, void 0, void 0, function* () {
        const document = context.getTextDocument(uri);
        if (!document)
            return;
        const offsetRange = {
            start: document.offsetAt(range.start),
            end: document.offsetAt(range.end),
        };
        let codeActions = yield (0, featureWorkers_1.languageFeatureWorker)(context, uri, { range, codeActionContext }, (arg, sourceMap) => {
            var _a;
            if (!sourceMap.embeddedFile.capabilities.codeActions)
                return [];
            const _codeActionContext = {
                diagnostics: (0, transforms_1.transformLocations)(codeActionContext.diagnostics, range => { var _a; return (_a = sourceMap.getMappedRange(range.start, range.end)) === null || _a === void 0 ? void 0 : _a[0]; }),
                only: codeActionContext.only,
            };
            let minStart;
            let maxEnd;
            for (const mapping of sourceMap.mappings) {
                const overlapRange = shared.getOverlapRange2(offsetRange, mapping.sourceRange);
                if (overlapRange) {
                    const embeddedRange = (_a = sourceMap.getMappedRange(overlapRange.start, overlapRange.end)) === null || _a === void 0 ? void 0 : _a[0];
                    if (embeddedRange) {
                        minStart = minStart === undefined ? embeddedRange.start : Math.min(embeddedRange.start, minStart);
                        maxEnd = maxEnd === undefined ? embeddedRange.end : Math.max(embeddedRange.end, maxEnd);
                    }
                }
            }
            if (minStart !== undefined && maxEnd !== undefined) {
                return [{
                        range: vscode.Range.create(sourceMap.mappedDocument.positionAt(minStart), sourceMap.mappedDocument.positionAt(maxEnd)),
                        codeActionContext: _codeActionContext,
                    }];
            }
            return [];
        }, (plugin, document, arg, sourceMap) => __awaiter(this, void 0, void 0, function* () {
            var _a, _b;
            const codeActions = yield ((_b = (_a = plugin.codeAction) === null || _a === void 0 ? void 0 : _a.on) === null || _b === void 0 ? void 0 : _b.call(_a, document, arg.range, arg.codeActionContext));
            return codeActions === null || codeActions === void 0 ? void 0 : codeActions.map(_codeAction => {
                const data = {
                    uri,
                    originalItem: _codeAction,
                    pluginId: context.getPluginId(plugin),
                    sourceMap: sourceMap ? {
                        embeddedDocumentUri: sourceMap.mappedDocument.uri,
                    } : undefined,
                };
                return Object.assign(Object.assign({}, _codeAction), { data: data });
            });
        }), (_codeActions, sourceMap) => _codeActions.map(_codeAction => {
            if (!sourceMap)
                return _codeAction;
            if (_codeAction.edit) {
                const edit = (0, rename_1.embeddedEditToSourceEdit)(_codeAction.edit, context.vueDocuments);
                if (edit) {
                    _codeAction.edit = edit;
                    return _codeAction;
                }
            }
            else {
                return _codeAction;
            }
        }).filter(shared.notEmpty), arr => arr.flat());
        if (codeActions) {
            codeActions = codeActions.filter(codeAction => codeAction.title.indexOf('__VLS_') === -1);
            return dedupe.withCodeAction(codeActions);
        }
    });
}
exports.register = register;
