// @volar/vue-language-service/out/languageFeatures/complete.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.register = void 0;
const transforms_1 = require("@volar/transforms");
const vscode = require("vscode-languageserver-protocol");
const definePlugin_1 = require("../utils/definePlugin");
function register(context) {
    let cache;
    return (uri, position, completionContext) => __awaiter(this, void 0, void 0, function* () {
        var _a, _b, _c, _d;
        let document;
        if ((completionContext === null || completionContext === void 0 ? void 0 : completionContext.triggerKind) === vscode.CompletionTriggerKind.TriggerForIncompleteCompletions
            && (cache === null || cache === void 0 ? void 0 : cache.uri) === uri) {
            for (const cacheData of cache.data) {
                if (!cacheData.list.isIncomplete)
                    continue;
                if (cacheData.sourceMap) {
                    const sourceMap = context.vueDocuments.sourceMapFromEmbeddedDocumentUri(cacheData.sourceMap.embeddedDocumentUri);
                    if (!sourceMap)
                        continue;
                    for (const [embeddedRange] of sourceMap.getMappedRanges(position, position, data => !!data.capabilities.completion)) {
                        if (!((_a = cacheData.plugin.complete) === null || _a === void 0 ? void 0 : _a.on))
                            continue;
                        const embeddedCompletionList = yield cacheData.plugin.complete.on(sourceMap.mappedDocument, embeddedRange.start, completionContext);
                        if (!embeddedCompletionList) {
                            cacheData.list.isIncomplete = false;
                            continue;
                        }
                        cacheData.list = Object.assign(Object.assign({}, embeddedCompletionList), { items: embeddedCompletionList.items.map(item => {
                                const data = {
                                    uri,
                                    originalItem: item,
                                    pluginId: context.getPluginId(cacheData.plugin),
                                    sourceMap: {
                                        embeddedDocumentUri: sourceMap.mappedDocument.uri,
                                    },
                                };
                                return Object.assign(Object.assign({}, (0, transforms_1.transformCompletionItem)(item, embeddedRange => { var _a; return (_a = sourceMap.getSourceRange(embeddedRange.start, embeddedRange.end)) === null || _a === void 0 ? void 0 : _a[0]; })), { data: data });
                            }) });
                    }
                }
                else if (document = context.getTextDocument(uri)) {
                    if (!((_b = cacheData.plugin.complete) === null || _b === void 0 ? void 0 : _b.on))
                        continue;
                    const completionList = yield cacheData.plugin.complete.on(document, position, completionContext);
                    if (!completionList) {
                        cacheData.list.isIncomplete = false;
                        continue;
                    }
                    cacheData.list = Object.assign(Object.assign({}, completionList), { items: completionList.items.map(item => {
                            const data = {
                                uri,
                                originalItem: item,
                                pluginId: context.getPluginId(cacheData.plugin),
                                sourceMap: undefined,
                            };
                            return Object.assign(Object.assign({}, item), { data: data });
                        }) });
                }
            }
        }
        else {
            const vueDocument = context.vueDocuments.get(uri);
            cache = {
                uri,
                data: [],
                mainCompletion: undefined,
            };
            // monky fix https://github.com/johnsoncodehk/volar/issues/1358
            let isFirstMapping = true;
            if (vueDocument) {
                const embeddeds = vueDocument.file.embeddeds;
                yield (0, definePlugin_1.visitEmbedded)(vueDocument, embeddeds, (sourceMap) => __awaiter(this, void 0, void 0, function* () {
                    var _e, _f;
                    const plugins = context.getPlugins().sort(sortPlugins);
                    for (const [embeddedRange, data] of sourceMap.getMappedRanges(position, position, data => !!data.capabilities.completion)) {
                        for (const plugin of plugins) {
                            if (!((_e = plugin.complete) === null || _e === void 0 ? void 0 : _e.on))
                                continue;
                            if (plugin.complete.isAdditional && !isFirstMapping)
                                continue;
                            if ((completionContext === null || completionContext === void 0 ? void 0 : completionContext.triggerCharacter) && !((_f = plugin.complete.triggerCharacters) === null || _f === void 0 ? void 0 : _f.includes(completionContext.triggerCharacter)))
                                continue;
                            const isAdditionalMapping = typeof data.capabilities.completion === 'object' && data.capabilities.completion.additional;
                            if (cache.mainCompletion && ((!plugin.complete.isAdditional && !isAdditionalMapping) || (cache === null || cache === void 0 ? void 0 : cache.mainCompletion.documentUri) !== sourceMap.mappedDocument.uri))
                                continue;
                            // avoid duplicate items with .vue and .vue.html
                            if (plugin.complete.isAdditional && (cache === null || cache === void 0 ? void 0 : cache.data.some(data => data.plugin === plugin)))
                                continue;
                            const embeddedCompletionList = yield plugin.complete.on(sourceMap.mappedDocument, embeddedRange.start, completionContext);
                            if (!embeddedCompletionList || !embeddedCompletionList.items.length)
                                continue;
                            if (!plugin.complete.isAdditional) {
                                cache.mainCompletion = { documentUri: sourceMap.mappedDocument.uri };
                            }
                            const completionList = Object.assign(Object.assign({}, embeddedCompletionList), { items: embeddedCompletionList.items.map(item => {
                                    const data = {
                                        uri,
                                        originalItem: item,
                                        pluginId: context.getPluginId(plugin),
                                        sourceMap: {
                                            embeddedDocumentUri: sourceMap.mappedDocument.uri,
                                        }
                                    };
                                    return Object.assign(Object.assign({}, (0, transforms_1.transformCompletionItem)(item, embeddedRange => { var _a; return (_a = sourceMap.getSourceRange(embeddedRange.start, embeddedRange.end)) === null || _a === void 0 ? void 0 : _a[0]; })), { data: data });
                                }) });
                            cache.data.push({
                                sourceMap: {
                                    embeddedDocumentUri: sourceMap.mappedDocument.uri,
                                },
                                plugin,
                                list: completionList,
                            });
                        }
                        isFirstMapping = false;
                    }
                    return true;
                }));
            }
            if (document = context.getTextDocument(uri)) {
                const plugins = context.getPlugins().sort(sortPlugins);
                for (const plugin of plugins) {
                    if (!((_c = plugin.complete) === null || _c === void 0 ? void 0 : _c.on))
                        continue;
                    if (plugin.complete.isAdditional && !isFirstMapping)
                        continue;
                    if ((completionContext === null || completionContext === void 0 ? void 0 : completionContext.triggerCharacter) && !((_d = plugin.complete.triggerCharacters) === null || _d === void 0 ? void 0 : _d.includes(completionContext.triggerCharacter)))
                        continue;
                    if (cache.mainCompletion && (!plugin.complete.isAdditional || cache.mainCompletion.documentUri !== document.uri))
                        continue;
                    // avoid duplicate items with .vue and .vue.html
                    if (plugin.complete.isAdditional && (cache === null || cache === void 0 ? void 0 : cache.data.some(data => data.plugin === plugin)))
                        continue;
                    const completionList = yield plugin.complete.on(document, position, completionContext);
                    if (!completionList || !completionList.items.length)
                        continue;
                    if (!plugin.complete.isAdditional) {
                        cache.mainCompletion = { documentUri: document.uri };
                    }
                    cache.data.push({
                        sourceMap: undefined,
                        plugin,
                        list: Object.assign(Object.assign({}, completionList), { items: completionList.items.map(item => {
                                const data = {
                                    uri,
                                    originalItem: item,
                                    pluginId: context.getPluginId(plugin),
                                    sourceMap: undefined,
                                };
                                return Object.assign(Object.assign({}, item), { data: data });
                            }) }),
                    });
                }
            }
        }
        return combineCompletionList(cache.data.map(cacheData => cacheData.list));
        function sortPlugins(a, b) {
            var _a, _b;
            return (((_a = b.complete) === null || _a === void 0 ? void 0 : _a.isAdditional) ? -1 : 1) - (((_b = a.complete) === null || _b === void 0 ? void 0 : _b.isAdditional) ? -1 : 1);
        }
        function combineCompletionList(lists) {
            return {
                isIncomplete: lists.some(list => list.isIncomplete),
                items: lists.map(list => list.items).flat().filter((result) => {
                    var _a;
                    return result.label.indexOf('__VLS_') === -1
                        && (!((_a = result.labelDetails) === null || _a === void 0 ? void 0 : _a.description) || result.labelDetails.description.indexOf('__VLS_') === -1);
                }),
            };
        }
    });
}
exports.register = register;
